### ISSUE TITLE
React 19 issue with useWindowVirtualizer

### ISSUE THREAD

**Issue by fat** [TanStack/virtual#743]
### Describe the bug

Sorry for the drive by issue… 

We ran into a problem with react 19 today that was a pita to track down. So dropping it here incase it's useful to anyone.

We had a simple useWindowVirutalizer in a client component in next.js w/ react 19 (like this):

```js
const virtualizer = useWindowVirtualizer({
		count: branches.length,
		estimateSize: () => 66,
		overscan: 5,
		scrollMargin: listRef.current?.offsetTop ?? 0,
	});
```

and virtualizerRef.getVirtualItems() was being hoisted / memo'd by the compiler in such a way that it was always returning 0 for `getVirutalItems`.

To fix it we just dropped the virtualizer in a ref…

```js
	const virtualizer = useWindowVirtualizer({
		count: branches.length,
		estimateSize: () => 66,
		overscan: 5,
		scrollMargin: listRef.current?.offsetTop ?? 0,
	});

	// we do this because react 19 was optimzing it away
	const virtualizerRef = useRef(virtualizer);
	
	…
	
	virtualizerRef.current.getVirtualItems() // etc…
```

### Your minimal, reproducible example

n/a

### Steps to reproduce

Don't have time to spin up a test case sorry! 

### Expected behavior

Expected getVirtualItems() to return items.

### How often does this bug happen?

Every time

### Platform

macos, chrome, react 19

### tanstack-virtual version

3.7.0

### Terms & Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I understand that if my bug cannot be reliable reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

**Comment by Prains** [comment]
same issue here, but droppind a virtualizer in a ref does not helps

**Comment by izakfilmalter** [comment]
Running into the same issue. `useRef` seems to work for now for us.

**Comment by eqqe** [comment]
Ut guves me an issue with eslint however.

```
Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)eslint(react-compiler/react-compiler)
```

**Comment by cacabo** [comment]
+1, I am seeing this with:

```json
{
    "@tanstack/react-virtual": "^3.11.1",
    "react": "^18.3.1",
    "react-compiler-runtime": "^19.0.0-beta-63b359f-20241101",
    "react-dom": "^18.3.1",
}
```

Putting a hook in the component after `.getVirutalItems` appears to be another workaround

**Comment by piecyk** [comment]
Yep, 'use no memo' seems like a reasonable temporary workaround! 👍

**Comment by zeorin** [comment]
I work around this with the following hook:

```typescript
const useNoMemo = <const T>(factory: () => T): T => {
  'use no memo';
  return factory();
}
```
which I then use like this:
```typescript
  const virtualItems = useNoMemo(() => virtualizer.getVirtualItems());
```
This de-opts _just the call to `virtualizer.getVirtualItems` without de-opting the rest of the component.

See https://github.com/facebook/react/issues/33057#issuecomment-2894450792

**Comment by kaushal-dhrangad-techno** [comment]
> I work around this with the following hook:
> 
> const useNoMemo = <const T>(factory: () => T): T => {
>   'use no memo';
>   return factory();
> }
> which I then use like this:
> 
>   const virtualItems = useNoMemo(() => virtualizer.getVirtualItems());
> This de-opts _just_ the call to `virtualizer.getVirtualItems` without de-opting the rest of the component.
> 
> See [facebook/react#33057 (comment)](https://github.com/facebook/react/issues/33057#issuecomment-2894450792)

+1 works for me :-) thanks!

**Comment by zeorin** [comment]
I've published a compiler-compatible package that is a near drop-in replacement for `@tanstack/react-virtual`: [react-hook-tanstack-virtual](https://www.npmjs.com/package/react-hook-tanstack-virtual).

**Comment by piecyk** [comment]
This has been addressed with the `directDomUpdates` option. When enabled, the virtualizer writes item positions and container sizes directly to the DOM, bypassing React's render cycle for scroll-driven updates. This makes the React Compiler's memoization of `getVirtualItems()` harmless, since position updates no longer depend on React re-renders.

```tsx
const virtualizer = useWindowVirtualizer({
  count: items.length,
  estimateSize: () => 66,
  overscan: 5,
  directDomUpdates: true,
});
```

The feature is covered by dedicated e2e tests that build with `babel-plugin-react-compiler` and verify scrolling works correctly.

Closing as resolved. If you're not using `directDomUpdates`, the `"use no memo"` workaround mentioned above remains a valid alternative.

**Comment by adi-works** [comment]
@piecyk Slick, but it didn't work for me. However, the `useNoMemo` workaround did work.
